File: data_structure_python/array/array.py
class Array:
    def __init__(self):
        self.data = []
        self.size = 0

    def __str__(self):
        res = '['
        for i in range(self.size):
            res += str(self.data[i])
            if i != self.size - 1:
                res += ', '
        res += ']'
        return res

    def __repr__(self):
        res = '['
        for i in range(self.size):
            res += repr(self.data[i])
            if i != self.size - 1:
                res += ', '
        res += ']'
        return res

    def __len__(self):
        return self.size

    def __getitem__(self, index):
        if index < 0 or index >= self.size:
            raise IndexError("index is out of range")
        return self.data[index]

    # return boolean if array is empty
    def is_empty(self):
        return self.size == 0

    # get element by index
    def get_element(self, index):
        if index < 0 or index > self.size:
            raise IndexError("Invalid Index")
        return self.data[index]

    # get first element
    def get_first(self):
        return self.get_element(0)

    # get last element
    def get_last(self):
        return self.get_element(self.size - 1)

    # find element by value, and return it's index or -1
    def find(self, value):
        for i in range(self.size):
            if self.data[i] == value:
                return i
        return -1

    # return boolean if array has element
    def contains(self, value):
        for i in range(self.size):
            if self.data[i] == value:
                return True
        return False

    # change element value by index
    def set(self, index, value):
        if index < 0 or index >= self.size:
            raise IndexError("Invalid Index")
        self.data[index] = value

    # add element by index
    def add(self, index, value):
        if index < 0 or index > self.size:
            raise IndexError("Invalid Index")
        self.data.insert(index, value)
        self.size += 1

    # add element on first of array
    def add_first(self, value):
        self.add(0, value)

    # add element on last of array
    def add_last(self, value):
        self.add(self.size, value)

    # remove element by index
    def remove(self, index):
        if index < 0 or index >= self.size:
            raise IndexError("Invalid Index")
        removed_element = self.data[index]
        for i in range(index + 1, self.size):
            self.data[i - 1] = self.data[i]
        self.size -= 1
        self.data[self.size] = None
        return removed_element

    # remove first element
    def remove_first(self):
        return self.remove(0)

    # remove last element
    def remove_last(self):
        return self.remove(self.size - 1)

    # delete element
    def remove_element(self, value):
        element_index = self.find(value)
        if element_index != -1:
            self.remove(element_index)

    # change 2 elements by index
    def swap(self, index1, index2):
        if index1 < 0 or index1 >= self.size or index2 < 0 or index2 >= self.size:
            temp = self.data[index1]
            self.data[index1] = self.data[index2]
            self.data[index2] = temp


'''if __name__ == '__main__':
    arr = Array()
    arr.add(0, 5)
    arr.add_last(3)
    arr.add_first(6)
    print(len(arr))
    print(arr)
    arr.remove_element(3)
    print(arr)'''

File: DataStructureJava/array/Array.java
public class Array <E>{
    private E[] data;
    private int size;

    public Array(int capacity){
        data = (E[]) new Object[capacity];
        size = 0;
    }

    public Array(){
        this(10);
    }

    public Array(E[] arr){
        data = (E[]) new Object[arr.length];
        for(int i = 0; i < arr.length; i ++)
            data[i] = arr[i];
        size = arr.length;

    }

    public int getSize(){
        return size;
    }

    public int getCapacity(){
        return data.length;
    }

    public boolean isEmpty(){
        return size == 0;
    }

    //get element by index
    public E get(int index){
        if (index<0 || index>size){
            throw new IllegalArgumentException("Get failed,Index is illegal");
        }
        return data[index];
    }

    public E getFirst(){
        return get(0);
    }

    public E getLast(){
        return get(size - 1);
    }

    //change element to new value by index
    public void set(int index,E e){
        if (index<0 || index>size){
            throw new IllegalArgumentException(("Get failed,Index is illegal"));
        }
      
        data[index] = e;
    }

    public boolean contains(E e){
        for (int i= 0;i < size;i ++){
            if(data[i].equals(e))
                return true;
        }
        return false;
    }
    //find element and return it's index.if element is not exist, return -1
    public int find(E e){
        for (int i= 0;i < size;i ++){
            if(data[i].equals(e))
                return i;
        }
        return -1;
    }


    // add element
    public void add(int index, E e){
        if (index < 0 || index > size){
            throw new IllegalArgumentException("Invalid Index");
        }
        if (size == data.length){
            resize( 2 * data.length);
        }
        for (int i = size-1; i >= index; i--){
            data[i+1] = data[i];
        }
        data[index] = e;
        size ++;
    }

    public void addLast(E e){
        add(size,e);
    }

    public void addFirst(E e){
        add(0,e);

    }

    //delete element on index???and return element
    public E remove(int index){
        if (index < 0 || index > size){
            throw new IllegalArgumentException("Invalid Index");
        }
        E ret = data[index];
        for(int i = index + 1;i < size;i++){
            data[i-1] = data[i];
        }
        size --;
        data[size] = null;
        if (size == data.length /2){
            resize(data.length / 2);
        }
        return ret;
    }
    public E removeFirst(){
        return remove(0);
    }
    public E removeLast(){
        return remove(size-1);
    }

    //delete the element
    public void removeElement(E e){
        int index = find(e);
        if(index != -1){
            remove(index);
        }
    }
    public void  swap(int i, int j){
        if (i < 0 || i >= size || j < 0 || j >= size){
            throw new IllegalArgumentException("Index is illegal");
        }
        E t = data[i];
        data[i] = data[j];
        data[j] = t;
    }

    @Override
    public String toString(){
        StringBuilder res = new StringBuilder();
        res.append(String.format("Array: size == %d , capacity == %d \n", size,data.length));
        res.append("[");
        for(int i=0;i <size;i++){
            res.append(data[i]);
            if(i != size - 1){
                res.append(",");
            }
        }
        res.append("]");
        return res.toString();
    }

    private void resize(int newCapacity){
        E[] newData = (E[]) new Object[newCapacity];
        for (int i = 0;i < size; i ++){
            newData[i] = data[i];
        }
        data = newData;
    }
}
